pkg/remotes/docker/artifacts.go

```diff
@@ -0,0 +1,266 @@
+package docker
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	remoteserrors "github.com/containerd/containerd/remotes/errors"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+func (d *dockerDiscoverer) Pusher(ctx context.Context, ref string) (remotes.Pusher, error) {
+	d.reference = ref
+	return d, nil
+}
+
+func (d *dockerDiscoverer) Push(ctx context.Context, desc ocispec.Descriptor) (content.Writer, error) {
+	switch desc.MediaType {
+	case artifactspec.MediaTypeArtifactManifest:
+		h, err := d.filterHosts(docker.HostCapabilityPush)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(h) == 0 {
+			return nil, errors.New("no host with push")
+		}
+
+		host := h[0]
+
+		err = d.CheckManifest(ctx, host, desc)
+		if err != nil {
+			return nil, err
+		}
+
+		return d.PreparePutManifest(ctx, host, desc)
+	}
+
+	pusher, err := d.Resolver.Pusher(ctx, d.reference)
+	if err != nil {
+		return nil, err
+	}
+
+	return pusher.Push(ctx, desc)
+}
+
+func (d *dockerDiscoverer) PreparePutManifest(ctx context.Context, host docker.RegistryHost, desc ocispec.Descriptor) (content.Writer, error) {
+	d.tracker.SetStatus(d.reference, docker.Status{
+		Status: content.Status{
+			Ref:       d.reference,
+			Total:     desc.Size,
+			Expected:  desc.Digest,
+			StartedAt: time.Now(),
+		},
+	})
+
+	req := d.request(host, http.MethodPut, "manifests", desc.Digest.String())
+	if req == nil {
+		return nil, errors.New("not implemented")
+	}
+
+	req.header.Set("Content-Type", desc.MediaType)
+
+	pr, pw := io.Pipe()
+	respC := make(chan response, 1)
+	body := ioutil.NopCloser(pr)
+
+	req.body = func() (io.ReadCloser, error) {
+		if body == nil {
+			return nil, errors.New("retry request, cannot reset the stream")
+		}
+
+		readout := body
+		body = nil
+		return readout, nil
+	}
+	req.size = desc.Size
+
+	go func() {
+		defer close(respC)
+		resp, err := req.doWithRetries(ctx, nil)
+		if err != nil {
+			respC <- response{err: err}
+			pr.CloseWithError(err)
+			return
+		}
+
+		switch resp.StatusCode {
+		case http.StatusOK, http.StatusCreated, http.StatusNoContent:
+		default:
+			err := remoteserrors.NewUnexpectedStatusErr(resp)
+			pr.CloseWithError(err)
+		}
+		respC <- response{Response: resp}
+	}()
+
+	return &artifactsManifest{
+		ref:       d.reference,
+		expected:  desc.Digest,
+		pipe:      pw,
+		responseC: respC,
```

Comment:
Why do you choose to return the channel? Will it be simple to get the content and return?

Reply:
This is exactly how the docker resolver already works. If you don't do it this way you lose the StatusTracker. And, I would rather not try to rewrite that particular part of their push code.